sdl2promela/stop_condition/parser.py

```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import antlr3
+import opengeode
+
+from opengeode import sdl92Lexer as lexer
+from .model import StopConditionModel
+from .model import Expression
+from .model import Selector
+from .model import IntegerValue
+from .model import FloatValue
+from .model import BooleanValue
+from .model import VariableReference
+from .model import CallExpression
+from .model import OrExpression
+from .model import AndExpression
+from .model import XorExpression
+from .model import EqualExpression
+from .model import NotEqualExpression
+from .model import LessExpression
+from .model import LessEqualExpression
+from .model import GreaterExpression
+from .model import GreaterEqualExpression
+from .model import NotExpression
+from .model import PlusExpression
+from .model import MinusExpression
+from .model import MulExpression
+from .model import DivExpression
+from .model import ModExpression
+from .model import NegationExpression
+from .model import AlwaysStatement
+from .model import NeverStatement
+from .model import EventuallyStatement
+from .model import FilterOutStatement
+
+
+class TranslationException(Exception):
+    """Exception during parsing Stop Condition model"""
+
+    pass
+
+
+def _parse_primary_expression(expr: antlr3.tree.CommonTree) -> Expression:
+    if expr.type == lexer.INT:
+        return IntegerValue(int(expr.getText()))
+    if expr.type == lexer.FLOAT:
+        return FloatValue(float(expr.getText()))
+    if expr.type == lexer.VARIABLE:
+        return VariableReference(expr.children[0].getText())
+    if expr.type == lexer.TRUE:
+        return BooleanValue(True)
+    if expr.type == lexer.FALSE:
+        return BooleanValue(False)
+    else:
+        raise TranslationException("Unexpected tree type: {}".format(expr.type))
+    return Expression()
+
+
+def _parse_selector(selector: antlr3.tree.CommonTree) -> Selector:
+    if len(selector.children) != 2:
+        raise TranslationException(
+            "Unexpected number of children in parse tree: expected 2 got {}".format(
+                len(selector.children)
+            )
+        )
+    if not selector.children[0].type in [lexer.SELECTOR, lexer.PRIMARY, lexer.CALL]:
+        raise TranslationException(
+            "Unexpected tree type {}".format(selector.children[0].type)
+        )
+    if selector.children[1].type != lexer.ID:
+        raise TranslationException(
+            "Unexpected tree type {}".format(selector.children[1].type)
+        )
+
+    elements: Selector.SelectorElements = []
+
+    if selector.children[0].type == lexer.SELECTOR:
+        nested = _parse_selector(selector.children[0])
+        elements = elements + nested.elements
+    elif selector.children[0].type == lexer.PRIMARY:
+        variable_reference = VariableReference(
+            selector.children[0].children[0].children[0].getText()
```

Comment:
Why not parse_primary?

Reply:
True. However, I would like to refactor this with automatic tests. I've added TODO